File: 0415-add-strings/0415-add-strings.cpp
class Solution {
public:
    string addStrings(string num1, string num2) {
        string result = "";
        int i = num1.size() - 1, j = num2.size() - 1, carry = 0;

        while (i >= 0 || j >= 0 || carry) {
            int digit1 = (i >= 0) ? (num1[i] - '0') : 0;
            int digit2 = (j >= 0) ? (num2[j] - '0') : 0;
            int sum = digit1 + digit2 + carry;
            
            result += (sum % 10) + '0';
            carry = sum / 10; 
            i--;
            j--;
        }

        reverse(result.begin(), result.end()); 
        return result;
    }
};

File: 0415-add-strings/0415-add-strings.py
class Solution(object):
    def addStrings(self, num1, num2):
        """
        :type num1: str
        :type num2: str
        :rtype: str
        """
        # return str(int(num1)+int(num2))
        result = ''
        i = len(num1) - 1
        j = len(num2) - 1
        carry = 0

        while i >= 0 or j >= 0 or carry:
            if i >= 0:
                digit1 = ord(num1[i]) - ord('0')
            else:
                digit1 = 0

            if j >= 0:
                digit2 = ord(num2[j]) - ord('0')
            else:
                digit2 = 0

            total = digit1 + digit2 + carry
            carry = total // 10
            digit = total % 10

            result += chr(digit + ord('0'))  # convert back to char
            i -= 1
            j -= 1

        # reverse the result
        res = ''
        for k in range(len(result)-1, -1, -1):
            res += result[k]

        return res
